### DISCUSSION TITLE
How exactly is `before-script-linux` expected to work to install dependencies?

### DISCUSSION THREAD

**Discussion by bluenote10** [messense/maturin-action#273] [Q&A] [answered]
https://github.com/PyO3/maturin-action/pull/157 has introduced `before-script-linux` with the goal to allow installing system dependencies inside the docker container.

I tried to use it via `before-script-linux: apt-get -y install <some-package>` but the CI pipeline failed with:

```
Run before script
  /home/runner/work/_temp/run-maturin-action.sh: line 21: apt-get: command not found
  Error: The process '/usr/bin/docker' failed with exit code 127
      at ExecState._setResult (/home/runner/work/_actions/PyO3/maturin-action/v1/dist/index.js:1702:25)
      at ExecState.CheckComplete (/home/runner/work/_actions/PyO3/maturin-action/v1/dist/index.js:1685:18)
      at ChildProcess.<anonymous> (/home/runner/work/_actions/PyO3/maturin-action/v1/dist/index.js:1579:27)
      at ChildProcess.emit (node:events:519:28)
      at maybeClose (node:internal/child_process:1105:16)
      at ChildProcess._handle.onexit (node:internal/child_process:305:5)
```

How exactly is `before-script-linux` expected to be used to install dependencies?

**Comment by bluenote10** [comment, marked as answer]
Trying to self-answer as far as possible.

The issue I was running into was basically a consequence of how the default generated `CI.yml` by maturin looks like: It contains a matrix over multiple Linux distributions which have different package managers. Therefore just using `apt-get` isn't sufficient, because it will fail in the matrix branches that don't have `apt-get`.

I've found a work-around in this comment: https://github.com/PyO3/maturin-action/discussions/162#discussioncomment-7978369

The idea is to introduce a branching into the `before-script-linux` so that it picks the right kind of package manager dependent on the matrix branch:

```yaml
          [...]
          before-script-linux: |
            case "${{ matrix.target }}" in
              "aarch64" | "armv7" | "s390x" | "ppc64le")
                # NOTE: pypa/manylinux docker images are Debian based
                sudo apt-get update
                sudo apt-get install -y pkg-config libssl-dev
                ;;
              "x86" | "x86_64")
                # NOTE: rust-cross/manylinux docker images are CentOS based
                yum update -y
                yum install -y openssl openssl-devel
                ;;
            esac
          [...]
```

**Reply by lmmx** [reply]
I came up with a similar approach but preferred to use specific entries per platform rather than a big if condition https://github.com/lmmx/polars-fastembed/blob/master/.github/workflows/CI.yml#L56
